File: backend/src/modules/Portfolio/Portfolio.Domain/ValueObjects/ProjectPeriod.ts
import { InvalidProjectPeriodException } from '../Exceptions/InvalidProjectPeriodException';
import { CalendarDate } from './CalendarDate';

export class ProjectPeriod {
  public constructor(public readonly startDate: CalendarDate, public readonly endDate: CalendarDate) {
    if (endDate.compareTo(startDate) < 0) {
      throw new InvalidProjectPeriodException(startDate.value, endDate.value);
    }
  }

  public get totalDays(): number { return this.startDate.daysUntil(this.endDate); }

  public equals(other: unknown): boolean {
    return other instanceof ProjectPeriod
      && this.startDate.equals(other.startDate)
      && this.endDate.equals(other.endDate);
  }
}

File: backend/modules/Portfolio/CromaticaCreativa.Modules.Portfolio.Domain/ValueObjects/ProjectPeriod.cs
using CromaticaCreativa.Modules.Portfolio.Domain.Exceptions;

namespace CromaticaCreativa.Modules.Portfolio.Domain.ValueObjects;

public sealed record ProjectPeriod
{
    public ProjectPeriod(DateOnly startDate, DateOnly endDate)
    {
        if (endDate < startDate)
        {
            throw new InvalidProjectPeriodException(startDate, endDate);
        }

        StartDate = startDate;
        EndDate = endDate;
    }

    public DateOnly StartDate { get; }

    public DateOnly EndDate { get; }

    public int TotalDays => EndDate.DayNumber - StartDate.DayNumber;
}
